File: 39. Combination Sum.java
class Solution {
    List<List<Integer>> ans;
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        ans = new ArrayList<>();
        if (candidates == null || candidates.length == 0) return ans;
        backtrack(0, target, candidates, new ArrayList<Integer>());
        return ans;
    }
    
    private void backtrack(int start, int required, int[] candidates, List<Integer> path) {
        if (required == 0) {
            ans.add(new ArrayList<>(path));
            return;
        } else if (required < 0) {
            return;
        }
        for (int i = start; i < candidates.length; i ++) {
            path.add(candidates[i]);
            backtrack(i, required - candidates[i], candidates, path);
            path.remove(path.size() - 1);
        }
    } 
}

File: 39. Combination Sum.py
class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        candidates.sort()
        res = []
        n = len(candidates)
        def backtrack(remain, comb, i):
            if remain < 0 or i == n:
                return
            if remain == 0:
                res.append(list(comb))
                return 
            comb.append(candidates[i])
            backtrack(remain - candidates[i], comb, i)
            comb.pop()
            backtrack(remain, comb, i + 1)
        backtrack(target, [], 0)
        return res